Login/Logout UI tests

KaylaCrush opened the pull request:
# Description

Adds login/logout UI tests using Cypress, for both mobile-size screens and desktop-size screens.

Partly Addresses #[1419]

## Type of change

- [ ] Note merging this changes the database configuration.
- [ ] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.org/developer/pr/) ideas
- [x] I have removed text in ( ) from the issue request
- [x] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.org/developer/cla/) and each author is listed in the Description section.

## Limitations

We still need a full suite of ui tests, this is just a small part of that.

KaylaCrush committed:
added login/logout and dropdown menu tests to the general ui tests

KaylaCrush committed:
Merge remote-tracking branch 'origin/development' into general-ui-tests

KaylaCrush committed:
generalized login/logout tests to both desktop and mobile viewports

KaylaCrush committed:
Merge branch 'development' into general-ui-tests

KaylaCrush committed:
Moved login/logout tests to separate file

KaylaCrush committed:
Moved login/logout tests to separate file

KaylaCrush committed:
Got desktop version of login/logout working properly

KaylaCrush committed:
got mobile tests working alsogit status

KaylaCrush committed:
Fixed syntax issues (spaces to tabs, added a couple missing semicolons)

KaylaCrush committed:
Added header to loginLogoutTest.cy.ts

KaylaCrush committed:
fix general_ui.cy.ts

re-commented out broken tests (hadn't meant to include this file in the commit)

huss commented:
Thanks to @KaylaCrush for another contribution. I've had a chance to look at the code and had a big picture question. src/cypress/e2e/general_ui.cy.ts has a number of changes where most of the code is still commented out. Can you help me understand why some previous code was removed and some new code was added? Thanks and I'll be carefully reading/testing the new code in src/cypress/e2e/loginLogoutTest.cy.ts soon but my overview reading found it seems fine.

KaylaCrush commented:
Ah, @huss I tried to only include one file in this PR, the new login logout file, but the general test file got pulled in early in the branch and I wasn't sure how to exclude it without deleting it. I was going to clean up the general ui tests next, sorry for the mess.

huss commented:
> Ah, @huss I tried to only include one file in this PR, the new login logout file, but the general test file got pulled in early in the branch and I wasn't sure how to exclude it without deleting it. I was going to clean up the general ui tests next, sorry for the mess.

@KaylaCrush Okay. If the changes stay then I can give any comments and then delay the merge until the next PR comes with the changes to the other file. That is okay with me. If you want to remove those changes then that can be done (and I should be able to help if needed).

huss commented on src/cypress/e2e/loginLogoutTest.cy.ts:
```diff
@@ -0,0 +1,100 @@
+			it('should allow a user to login and logout successfully',()=>{
+
+				// On mobile, we have to click the menu button first
+				if(name == 'mobile'){
```
The formatting is not what OED prefers. If you are using VSC then the command ``format document`` should fix this up. Please let me know if I can help.

huss commented on src/cypress/e2e/loginLogoutTest.cy.ts:
```diff
@@ -0,0 +1,100 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+function getOptionsButton() {
```
While these functions are fairly clear as to what they do, comments are always nice.

As the UI tests are done, I'm wondering about using these functions across files and possibly generalizing them. This does not have to happen now but noting.

huss commented on src/cypress/e2e/loginLogoutTest.cy.ts:
```diff
@@ -0,0 +1,100 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+function getOptionsButton() {
+	return cy.get('a.dropdown-toggle:contains("Options")').last();
```
For all the strings used, I'm wondering if the translation key could be used. This would mean that if the exact text is changed via the key then the test would still work. The keys are all in src/client/app/translations/data.ts. For example, "Options" is key "option" and "Log in" is key "log.in". This would also make it easier, if desired at some point, to test other language choices. The ``translate`` function can convert the key to the expected text string.

huss reviewed (changes requested):
Thanks to @KaylaCrush for this contribution. Review and testing found it works as desired. I made a few comments to consider. Please let me know if anything is not clear or I can help.

KaylaCrush committed:
Added language translation support

KaylaCrush commented:
@huss ok, I figured out translations, fixed the formatting, and added some comments for more context. Let me know if there's anything else I can do! And thanks so much for helping me with this process.

huss commented:
> @KaylaCrush Okay. If the changes stay then I can give any comments and then delay the merge until the next PR comes with the changes to the other file. That is okay with me. If you want to remove those changes then that can be done (and I should be able to help if needed).

I looked at this more carefully and the old code seems to be there. The significant changes that appear seem to be how Git is showing the changes. Given this, I'm not holding up this PR for the next step.

huss reviewed (approved):
Thanks to @KaylaCrush for addressing all my comments. Review and testing found everything looks good. Congratulations on another accepted contribution.
